src/server/services/wmts/qgswmtsutils.cpp

```diff
@@ -0,0 +1,515 @@
+/***************************************************************************
+                              qgswmtsutils.cpp
+                              -------------------------
+  begin                : July 23, 2018
+  copyright            : (C) 2018 by René-Luc D'Hont
+  email                : rldhont at 3liz dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                  *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgswmtsutils.h"
+#include "qgsconfigcache.h"
+#include "qgsserverprojectutils.h"
+
+#include "qgsproject.h"
+#include "qgsexception.h"
+#include "qgsmapserviceexception.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgslayertree.h"
+#include "qgslayertreemodel.h"
+#include "qgslayertreemodellegendnode.h"
+#include "qgssettings.h"
+
+
+namespace QgsWmts
+{
+  namespace
+  {
+    QMap< QgsUnitTypes::DistanceUnit, double> populateInchesPerUnit();
+    QMap< QString, tileMatrixInfo> populateTileMatrixInfoMap();
+
+    QgsCoordinateReferenceSystem wgs84 = QgsCoordinateReferenceSystem::fromOgcWmsCrs( GEO_EPSG_CRS_AUTHID );
+
+    int DOTS_PER_INCH = 72;
+    double METERS_PER_INCH = 0.02540005080010160020;
+    QMap< QgsUnitTypes::DistanceUnit, double> INCHES_PER_UNIT = populateInchesPerUnit();
+    int tileWidth = 256;
+    int tileHeight = 256;
+
+    QMap< QString, tileMatrixInfo> tileMatrixInfoMap = populateTileMatrixInfoMap();
+  }
+
+  QString implementationVersion()
+  {
+    return QStringLiteral( "1.0.0" );
+  }
+
+
+  QString serviceUrl( const QgsServerRequest &request, const QgsProject *project )
+  {
+    QString href;
+    if ( project )
+    {
+      href = QgsServerProjectUtils::wmsServiceUrl( *project );
+    }
+
+    // Build default url
+    if ( href.isEmpty() )
+    {
+      QUrl url = request.url();
+      QUrlQuery q( url );
+
+      q.removeAllQueryItems( QStringLiteral( "REQUEST" ) );
+      q.removeAllQueryItems( QStringLiteral( "VERSION" ) );
+      q.removeAllQueryItems( QStringLiteral( "SERVICE" ) );
+      q.removeAllQueryItems( QStringLiteral( "_DC" ) );
```

Comment:
Same comment about `QgsServerParameters`. Enum should be used instead of string. For example:

```
params.remove( QgsServerParameter::REQUEST );
```

There was an error in this particular case with WFS services due to capitalization errors.

BTW, what is the `_DC` item?